File: docs/examples/python/hermes_client.py
#!/usr/bin/python

#
# Copyright (c) 2017 Cossack Labs Limited
#
# This file is a part of Hermes-core.
#
# Hermes-core is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Hermes-core is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with Hermes-core.  If not, see <http://www.gnu.org/licenses/>.
#
#

import socket
import argparse
import base64
import json
import hermes


class Transport(object):
    def __init__(self, host, port):
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.socket.connect((host, port))

    def __del__(self):
        if self.socket:
            self.socket.close()

    def send(self, msg):
        total_sent = 0
        message_length = len(msg)
        while total_sent < message_length:
            sent = self.socket.send(msg[total_sent:])
            if not sent:
                raise RuntimeError("socket connection broken")
            total_sent = total_sent + sent

    def receive(self, needed_length):
        chunks = []
        bytes_recieved = 0
        while bytes_recieved < needed_length:
            chunk = self.socket.recv(needed_length - bytes_recieved)
            if not chunk:
                raise RuntimeError("socket connection broken")
            chunks.append(chunk)
            bytes_recieved = bytes_recieved + len(chunk)
        return b''.join(chunks)


parser = argparse.ArgumentParser(description='Hermes client example.')
parser.add_argument('--id', dest='id', required=True, help='user identificator')
parser.add_argument('--private_key', '-pk', dest='private_key', required=True,
                    help='path to private key)')

parser.add_argument('--add', '-a', action='store_true', default=False,
                    dest='add')
parser.add_argument('--read', '-r', action='store_true', default=False,
                    dest='read')
parser.add_argument('--update', '-u', action='store_true', default=False,
                    dest='update')
parser.add_argument('--delete', '-d', action='store_true', default=False,
                    dest='delete')

parser.add_argument('--rotate', '-rt', action='store_true', default=False, dest='rotate')

parser.add_argument('--grant_read', '-gr', action='store_true', default=False,
                    dest='grant_read')
parser.add_argument('--grant_update', '-gu', action='store_true', default=False,
                    dest='grant_update')
parser.add_argument('--revoke_read', '-rr', action='store_true', default=False,
                    dest='revoke_read')
parser.add_argument('--revoke_update', '-ru', action='store_true', default=False,
                    dest='revoke_update')


parser.add_argument('--doc', dest='doc_file_name', required=True,
                    help='document file name')
parser.add_argument('--meta', dest='meta', help='document meta data')
parser.add_argument('--for_user', dest='for_user',
                    help='peer user identifier')


parser.add_argument('--config', dest='config', help='Path to config file', default='config.json')

args = parser.parse_args()

with open(args.private_key, 'rb') as f:
    private_key = f.read()

with open(args.config, 'r') as f:
    config = json.load(f)

credential_store_transport = Transport(config['credential_store_host'], config['credential_store_port'])
key_store_transport = Transport(config['key_store_host'], config['key_store_port'])
data_store_transport = Transport(config['data_store_host'], config['data_store_port'])

credential_store_secure_transport = hermes.SecureHermesTransport(
    args.id.encode('utf-8'), private_key, config['credential_store_id'].encode("utf-8"),
    base64.b64decode(config['credential_store_public_key'].encode('utf-8')), credential_store_transport, False)
key_store_secure_transport = hermes.SecureHermesTransport(
    args.id.encode('utf-8'), private_key, config['key_store_id'].encode("utf-8"),
    base64.b64decode(config['key_store_public_key'].encode('utf-8')), key_store_transport, False)
data_store_secure_transport = hermes.SecureHermesTransport(
    args.id.encode('utf-8'), private_key, config['data_store_id'].encode("utf-8"),
    base64.b64decode(config['data_store_public_key'].encode('utf-8')), data_store_transport, False)


mid_hermes = hermes.MidHermes(
    args.id, private_key, credential_store_secure_transport,
    data_store_secure_transport, key_store_secure_transport)

if not (args.add or args.read or args.update or args.delete or args.rotate or
            args.grant_read or args.grant_update or args.revoke_update or
            args.revoke_read):
    print("choose any command add|read|update|delete|rotate|grant_read|grant_update|"
          "revoke_read|revoke_update")
    exit(1)


if args.add and args.meta is not None:
    block = open(args.doc_file_name, 'rb').read()
    mid_hermes.addBlock(args.doc_file_name.encode(), block, args.meta.encode())
    print('added <{}> with meta <{}>'.format(args.doc_file_name, args.meta))
elif args.read:
    print(mid_hermes.getBlock(args.doc_file_name.encode()))
elif args.update:
    block = open(args.doc_file_name, 'rb').read()
    mid_hermes.updBlock(args.doc_file_name.encode(), block, args.meta.encode())
    print('updated <{}> with meta <{}>'.format(args.doc_file_name, args.meta))
elif args.delete:
    mid_hermes.delBlock(args.doc_file_name.encode())
    print('deleted <{}>'.format(args.doc_file_name))
elif args.rotate:
    mid_hermes.rotateBlock(args.doc_file_name.encode())
    print('rotated <{}>'.format(args.doc_file_name))
elif args.grant_read:
    mid_hermes.grantReadAccess(
        args.doc_file_name.encode(), args.for_user.encode())
    print('granted read access <{}> for user <{}>'.format(
        args.doc_file_name, args.for_user))
elif args.grant_update:
    mid_hermes.grantUpdateAccess(
        args.doc_file_name.encode(), args.for_user.encode())
    print('granted update access <{}> for user <{}>'.format(
        args.doc_file_name, args.for_user))
elif args.revoke_read:
    mid_hermes.denyReadAccess(
        args.doc_file_name.encode(), args.for_user.encode())
    print('revoked read access <{}> for user <{}>'.format(
        args.doc_file_name, args.for_user))
elif args.revoke_update:
    mid_hermes.denyUpdateAccess(
        args.doc_file_name.encode(), args.for_user.encode())
    print('revoked update access <{}> for user <{}>'.format(
        args.doc_file_name, args.for_user))

print('done')

File: docs/examples/go/hermes_client.go
//
// Copyright (c) 2017 Cossack Labs Limited
//
// This file is a part of Hermes-core.
//
// Hermes-core is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Hermes-core is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Hermes-core.  If not, see <http://www.gnu.org/licenses/>.
//
//

package main

import (
	"net"
	"github.com/cossacklabs/hermes-core/gohermes"
	"github.com/cossacklabs/hermes-core/gohermes/midhermes/rpc"
	"encoding/base64"
	"flag"
	"fmt"
	"io/ioutil"
	"github.com/BurntSushi/toml"
)

type TCPTransport struct {
	connection net.Conn
}

func NewTransport(uri string) (TCPTransport, error) {
	conn, err := net.Dial("tcp", uri)
	if err != nil {
		return TCPTransport{}, err
	}
	return TCPTransport{conn}, nil
}

func (t TCPTransport) Write(buf []byte) error {
	var total_send = 0
	for total_send < len(buf) {
		sended, err := t.connection.Write(buf[total_send:])
		if nil != err {
			return err
		}
		total_send += sended
	}
	return nil
}

func (t TCPTransport) Read(buf []byte) error {
	var total_received = 0
	for total_received < len(buf) {
		received, err := t.connection.Read(buf[total_received:])
		if nil != err {
			return err
		}
		total_received += received
	}
	return nil
}

func (t TCPTransport) Close() error {
	return t.connection.Close()
}

type Config struct {
	CredentialStoreUrl       string `toml:"credential_store_url"`
	CredentialStoreId        string `toml:"credential_store_id"`
	CredentialStorePublicKey string `toml:"credential_store_public_key"`

	DataStoreUrl       string `toml:"data_store_url"`
	DataStoreId        string `toml:"data_store_id"`
	DataStorePublicKey string `toml:"data_store_public_key"`

	KeyStoreUrl       string `toml:"key_store_url"`
	KeyStoreId        string `toml:"key_store_id"`
	KeyStorePublicKey string `toml:"key_store_public_key"`
}

func usage() {
	fmt.Println(`hermes_client:
  -command string
    	command [add_block | read_block | update_block | delete_block | rotate_block | grant_read_access | grant_update_access | revoke_read_access | revoke_update_access ]
  -doc string
    	doc
  -for_user string
    	for user id
  -meta string
    	meta`)
}

func main() {
	var config = flag.String("config", "client.conf", "Config with settings")
	var userId = flag.String("id", "", "User's id that run command")
	var userPrivateKey = flag.String("private_key", "", "User's private key")
	var docFileName = flag.String("doc", "", "doc")
	var meta = flag.String("meta", "", "meta")
	var forUser = flag.String("for_user", "", "for user id")
	var command = flag.String("command", "", "command [add_block | read_block | update_block | delete_block | rotate_block | grant_read_access | grant_update_access | revoke_read_access | revoke_update_access ]")
	flag.Parse()
	var conf Config
	if _, err := toml.DecodeFile(*config, &conf); err != nil {
		panic(err)
	}
	flag.Usage = usage

	if *userId == "" || *userPrivateKey == "" ||
		conf.CredentialStoreId == "" || conf.CredentialStorePublicKey == "" || conf.CredentialStoreUrl == "" ||
		conf.KeyStoreUrl == "" || conf.KeyStoreId == "" || conf.KeyStorePublicKey == "" ||
		conf.DataStoreUrl == "" || conf.DataStoreId == "" || conf.DataStorePublicKey == "" {
		fmt.Println("all parameters from config file are required")
		flag.Usage()
		return
	}
	if *docFileName == "" || *command == "" {
		fmt.Println("<command> and <doc> parameters are required")
	}
	privateKey, err := base64.StdEncoding.DecodeString(*userPrivateKey)
	if nil != err {
		panic(err)
		return
	}
	CredentialStoreTransport, err := NewTransport(conf.CredentialStoreUrl)
	if nil != err {
		panic(err)
		return
	}
	credentialPublic, err := base64.StdEncoding.DecodeString(conf.CredentialStorePublicKey)
	if err != nil {
		panic(err)
	}
	defer CredentialStoreTransport.Close()
	secureCredentialTransport, err := rpc.NewSecureTransport([]byte(*userId), privateKey, []byte(conf.CredentialStoreId), credentialPublic, CredentialStoreTransport, false)
	if err != nil {
		panic(err)
	}

	dataStorePublic, err := base64.StdEncoding.DecodeString(conf.DataStorePublicKey)
	if err != nil {
		panic(err)
	}
	DataStoreTransport, err := NewTransport(conf.DataStoreUrl)
	if nil != err {
		panic(err)
		return
	}
	defer DataStoreTransport.Close()
	secureDataStoreTransport, err := rpc.NewSecureTransport([]byte(*userId), privateKey, []byte(conf.DataStoreId), dataStorePublic, DataStoreTransport, false)
	if err != nil {
		panic(err)
	}

	keyPublic, err := base64.StdEncoding.DecodeString(conf.KeyStorePublicKey)
	if err != nil {
		panic(err)
	}
	KeyStoreTransport, err := NewTransport(conf.KeyStoreUrl)
	if nil != err {
		panic(err)
	}
	defer KeyStoreTransport.Close()
	secureKeyStoreTransport, err := rpc.NewSecureTransport([]byte(*userId), privateKey, []byte(conf.KeyStoreId), keyPublic, KeyStoreTransport, false)
	if err != nil {
		panic(err)
	}
	rpcHermes, err := rpc.NewRPCMidHermes([]byte(*userId), privateKey, secureCredentialTransport, secureDataStoreTransport, secureKeyStoreTransport)
	if err != nil{
		panic(err)
	}
	mid_hermes, err := gohermes.NewMidHermes(rpcHermes)
	if nil != err {
		panic(err)
		return
	}
	defer mid_hermes.Close()

	switch *command {
	case "add_block":
		if *meta == "" {
			flag.Usage()
		}
		data, err := ioutil.ReadFile(*docFileName)
		if nil != err {
			panic(err)
		}
		err = mid_hermes.AddBlock([]byte(*docFileName), data, []byte(*meta))
		if nil != err {
			panic(err)
		}
	case "read_block":
		data, meta, err := mid_hermes.ReadBlock([]byte(*docFileName))
		if nil != err {
			panic(err)
		}
		fmt.Println(string(data))
		fmt.Println(string(meta))
	case "update_block":
		if *meta == "" {
			flag.Usage()
		}
		data, err := ioutil.ReadFile(*docFileName)
		if nil != err {
			panic(err)
		}
		err = mid_hermes.UpdateBlock([]byte(*docFileName), data, []byte(*meta))
		if nil != err {
			panic(err)
		}
	case "delete_block":
		err := mid_hermes.DeleteBlock([]byte(*docFileName))
		if nil != err {
			panic(err)
		}
	case "rotate_block":
		err := mid_hermes.RotateBlock([]byte(*docFileName))
		if nil != err {
			panic(err)
		}
	case "grant_read_access":
		if *forUser == "" {
			flag.Usage()
			return
		}
		err := mid_hermes.GrantReadAccess([]byte(*docFileName), []byte(*forUser))
		if nil != err {
			panic(err)
		}
	case "grant_update_access":
		if *forUser == "" {
			flag.Usage()
			return
		}
		err := mid_hermes.GrantWriteAccess([]byte(*docFileName), []byte(*forUser))
		if nil != err {
			panic(err)
		}
	case "revoke_read_access":
		if *forUser == "" {
			flag.Usage()
			return
		}
		err := mid_hermes.RevokeReadAccess([]byte(*docFileName), []byte(*forUser))
		if nil != err {
			panic(err)
		}
	case "revoke_update_access":
		if *forUser == "" {
			flag.Usage()
			return
		}
		err := mid_hermes.RevokeWriteAccess([]byte(*docFileName), []byte(*forUser))
		if nil != err {
			panic(err)
		}
	default:
		flag.Usage()
	}
	fmt.Println("success")
}
